openmmtools/mcmc.py

```diff
@@ -162,10 +166,62 @@ def apply(self, thermodynamic_state, sampler_state):
         sampler_state : openmmtools.states.SamplerState
            The initial sampler state before applying the move. This may
            be modified depending on the implementation.
-
+        context_cache : opemmtools.cache.ContextCache
+            Context cache to be used in the propagation of the mcmc move. If None,
+            it will use the global context cache.
         """
         pass
 
+    @property
+    def context_cache(self):
+        logger.warning('The MCMCMove.context_cache field has been deprecated. The API now requires context_cache '
+                       'be passed to apply(). Please update your code.')
+        from openmmtools import cache
+        return cache.global_context_cache
+
+    @staticmethod
+    def _get_context_cache(context_cache_input):
+        """
+        Method to return context to be used for move propagation.
+
+        .. note:: centralized API point to deal with context cache behavior.
+
+        Parameters
+        ----------
+        context_cache_input : openmmtools.cache.ContextCache or None
+            Context cache to be used in the propagation of the mcmc move. If None,
+            it will create a new unlimited ContextCache object.
+
+        Returns
+        -------
+        context_cache : openmmtools.cache.ContextCache
+            Context cache object to be used for propagation.
+        """
+        if context_cache_input is None:
+            # Default behavior, gobal Context Cache
+            context_cache = cache.global_context_cache
+        elif isinstance(context_cache_input, cache.ContextCache):
+            context_cache = context_cache_input
+        else:
+            raise ValueError("Context cache input is not a valid ContextCache or None type.")
+        return context_cache
+
+    def __deepcopy__(self, memo):
+        """Overriding default deep copy behavior to avoid creating new ContextCache objects without need."""
```

Comment:
You are right, now that the `context_cache` property defaults to the global context cache we shouldn't be needing this,

Reply:
We still need it, because `MultiStateSampler` makes a deepcopy of the `mcmc_moves`, so we don't want each copy to have its own `context_cache` because it will end up eating memory without need (issue in #521 )

Reply:
Can you elaborate a bit more?`MCMCMove` and derived classes no longer contain a `context_cache` attribute.

Reply:
Oh, you are correct! it's not a property and this one shouldn't be affected by the deepcopying. I agree, I don't think we need it.

Reply:
Changed in current state of the PR.